File: common/src/main/java/com/bondhub/common/security/SecurityContextFilter.java
package com.bondhub.common.security;

import com.bondhub.common.enums.Role;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Component
@Slf4j
@ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.SERVLET)
public class SecurityContextFilter extends OncePerRequestFilter {

    private static final String HEADER_ACCOUNT_ID = "X-Account-Id";
    private static final String HEADER_USER_ID = "X-User-Id";
    private static final String HEADER_USER_EMAIL = "X-User-Email";
    private static final String HEADER_USER_ROLES = "X-User-Roles";
    private static final String HEADER_USER_JTI = "X-JWT-Id";
    private static final String HEADER_REMAINING_TTL = "X-Remaining-TTL";

    @Override
    protected void doFilterInternal(HttpServletRequest request,
            HttpServletResponse response,
            FilterChain filterChain) throws ServletException, IOException {

        String accountId = request.getHeader(HEADER_ACCOUNT_ID);
        String userId = request.getHeader(HEADER_USER_ID);
        String email = request.getHeader(HEADER_USER_EMAIL);
        String rolesHeader = request.getHeader(HEADER_USER_ROLES);
        String jti = request.getHeader(HEADER_USER_JTI);
        Long remainingTTL = request.getHeader(HEADER_REMAINING_TTL) != null
                ? Long.parseLong(request.getHeader(HEADER_REMAINING_TTL))
                : null;

        if (accountId != null && email != null) {
            try {
                List<GrantedAuthority> authorities = parseRoles(rolesHeader);

                UserPrincipal userPrincipal = new UserPrincipal(accountId, userId, email, jti, remainingTTL,
                        authorities);

                UsernamePasswordAuthenticationToken authentication = new UsernamePasswordAuthenticationToken(
                        userPrincipal,
                        null,
                        authorities);

                SecurityContextHolder.getContext().setAuthentication(authentication);

                log.debug("Security context set for user: {} ({}) and account: {}", email, userId, accountId);

            } catch (Exception e) {
                log.error("Error setting security context: {}", e.getMessage());
            }
        }

        try {
            filterChain.doFilter(request, response);
        } finally {
            SecurityContextHolder.clearContext();
        }
    }

    private List<GrantedAuthority> parseRoles(String rolesHeader) {
        if (rolesHeader == null || rolesHeader.trim().isEmpty()) {
            return List.of(new SimpleGrantedAuthority("ROLE_USER"));
        }

        return Arrays.stream(rolesHeader.split(","))
                .map(String::trim)
                .filter(role -> !role.isEmpty())
                .map(role -> {
                    String roleWithPrefix = role.startsWith("ROLE_") ? role : "ROLE_" + role;
                    return new SimpleGrantedAuthority(roleWithPrefix);
                })
                .collect(Collectors.toList());
    }
}

File: post-recommendation-service/app/security/security_context_filter.py
import logging

from fastapi import Request
from starlette.middleware.base import BaseHTTPMiddleware

from app.security.principal import UserPrincipal
from app.security.security_context import security_principal_ctx

logger = logging.getLogger(__name__)


class SecurityContextFilter(BaseHTTPMiddleware):
    HEADER_ACCOUNT_ID = "X-Account-Id"
    HEADER_USER_ID = "X-User-Id"
    HEADER_USER_EMAIL = "X-User-Email"
    HEADER_USER_ROLES = "X-User-Roles"
    HEADER_USER_JTI = "X-JWT-Id"
    HEADER_REMAINING_TTL = "X-Remaining-TTL"

    @staticmethod
    def _parse_roles(raw_roles: str | None) -> list[str]:
        if not raw_roles or not raw_roles.strip():
            return ["ROLE_USER"]

        roles: list[str] = []
        for role in raw_roles.split(","):
            normalized = role.strip()
            if not normalized:
                continue
            roles.append(normalized if normalized.startswith("ROLE_") else f"ROLE_{normalized}")
        return roles or ["ROLE_USER"]

    async def dispatch(self, request: Request, call_next):
        token = None
        request.state.user_principal = None

        account_id = request.headers.get(self.HEADER_ACCOUNT_ID)
        email = request.headers.get(self.HEADER_USER_EMAIL)
        if account_id and email:
            try:
                remaining_ttl_raw = request.headers.get(self.HEADER_REMAINING_TTL)
                remaining_ttl = int(remaining_ttl_raw) if remaining_ttl_raw else None

                principal = UserPrincipal(
                    account_id=account_id,
                    user_id=request.headers.get(self.HEADER_USER_ID),
                    email=email,
                    jti=request.headers.get(self.HEADER_USER_JTI),
                    remaining_ttl=remaining_ttl,
                    roles=self._parse_roles(request.headers.get(self.HEADER_USER_ROLES)),
                )
                request.state.user_principal = principal
                token = security_principal_ctx.set(principal)
                logger.debug(
                    "Security context set for email=%s user_id=%s account_id=%s",
                    principal.email,
                    principal.user_id,
                    principal.account_id,
                )
            except Exception:
                logger.exception("Error setting security context")

        try:
            return await call_next(request)
        finally:
            if token is not None:
                security_principal_ctx.reset(token)
